File: gaiag/baseline/Twotopon.java
// Dezyne --- Dezyne command line tools
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

class Twotopon extends Component {

  Boolean b;

  ITwotopon i;

  public Twotopon(Runtime runtime) {this(runtime, "");};

  public Twotopon(Runtime runtime, String name) {this(runtime, name, null);};

  public Twotopon(Runtime runtime, String name, SystemComponent parent) {
    super(runtime, name, parent);
    this.flushes = true;
    b = false;
    i = new ITwotopon();
    i.in.name = "i";
    i.in.self = this;
    i.in.e = new Action() {public void action() {Runtime.callIn(Twotopon.this, new Action() {public void action() {i_e();}}, new Meta(Twotopon.this.i, "e"));};};

    i.in.t = new Action() {public void action() {Runtime.callIn(Twotopon.this, new Action() {public void action() {i_t();}}, new Meta(Twotopon.this.i, "t"));};};

  };
  public void i_e() {
    if (b) {
      i.out.a.action();
    }
    else if (! (b)) {
      i.out.a.action();
    }
  };

  public void i_t() {
    i.out.a.action();
  };

}

File: gaiag/baseline/Twotopon.cc
// Dezyne --- Dezyne command line tools
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
// Copyright © 2015 Rutger van Beusekom <rutger.van.beusekom@verum.com>
// Copyright © 2015 Paul Hoogendijk <paul.hoogendijk@verum.com>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include "Twotopon.hh"

#include "locator.hh"
#include "runtime.hh"

#include <iostream>

namespace dezyne
{
  Twotopon::Twotopon(const locator& dezyne_locator)
  : dzn_meta{"","Twotopon",reinterpret_cast<const component*>(this),0,{},{[this]{i.check_bindings();}}}
  , dzn_rt(dezyne_locator.get<runtime>())
  , b(false)
  , i({{"i",this},{"",0}})
  {
    dzn_rt.performs_flush(this) = true; 
    i.in.e = [&] () {
      call_in(this, [this] {i_e();}, std::make_tuple(&i, "e", "return"));
    };
    i.in.t = [&] () {
      call_in(this, [this] {i_t();}, std::make_tuple(&i, "t", "return"));
    };
  }

  void Twotopon::i_e()
  {
    if (b)
    {
      i.out.a();
    }
    else if (not (b))
    {
      i.out.a();
    }
  }

  void Twotopon::i_t()
  {
    i.out.a();
  }


}
